mmaction/core/evaluation/eval_detection.py

```diff
@@ -0,0 +1,249 @@
+        ap = np.zeros((len(self.tiou_thresholds), len(self.activity_index)))
+
+        # Adaptation to query faster
+        ground_truth_by_label = self.ground_truth.groupby('label')
+        prediction_by_label = self.prediction.groupby('label')
+
+        results = Parallel(n_jobs=len(self.activity_index))(
+            delayed(compute_average_precision_detection)(
+                ground_truth=ground_truth_by_label.get_group(cidx).reset_index(
+                    drop=True),
+                prediction=self._get_predictions_with_label(
+                    prediction_by_label, label_name, cidx),
+                tiou_thresholds=self.tiou_thresholds,
+            ) for label_name, cidx in list(self.activity_index.items()))
+
+        for i, cidx in enumerate(self.activity_index.values()):
+            ap[:, cidx] = results[i]
+
+        return ap
+
+    def evaluate(self):
+        """Evaluates a prediction file.
+
+        For the detection task we measure the interpolated mean average
+        precision to measure the performance of a method.
+        """
+        self.ap = self.wrapper_compute_average_precision()
+
+        self.mAP = self.ap.mean(axis=1)
+        self.average_mAP = self.mAP.mean()
+
+        return self.mAP, self.average_mAP
+
+
+def compute_average_precision_detection(ground_truth,
+                                        prediction,
+                                        tiou_thresholds=np.linspace(
+                                            0.5, 0.95, 10)):
+    """Compute average precision (detection task) between ground truth and
+    predictions data frames. If multiple predictions occurs for the same
+    predicted segment, only the one with highest score is matches as true
+    positive. This code is greatly inspired by Pascal VOC devkit.
+
+    Parameters
+    ----------
+    ground_truth : df
+        Data frame containing the ground truth instances.
+        Required fields: ['video-id', 't-start', 't-end']
+    prediction : df
+        Data frame containing the prediction instances.
+        Required fields: ['video-id, 't-start', 't-end', 'score']
+    tiou_thresholds : 1darray, optional
+        Temporal intersection over union threshold.
+
+    Outputs
+    -------
+    ap : float
+        Average precision score.
+    """
+    ap = np.zeros(len(tiou_thresholds))
+    if prediction.empty:
+        return ap
+
+    npos = float(len(ground_truth))
+    lock_gt = np.ones((len(tiou_thresholds), len(ground_truth))) * -1
+    # Sort predictions by decreasing score order.
+    sort_idx = prediction['score'].values.argsort()[::-1]
+    prediction = prediction.loc[sort_idx].reset_index(drop=True)
+
+    # Initialize true positive and false positive vectors.
+    tp = np.zeros((len(tiou_thresholds), len(prediction)))
+    fp = np.zeros((len(tiou_thresholds), len(prediction)))
+
+    # Adaptation to query faster
+    ground_truth_gbvn = ground_truth.groupby('video-id')
```

Comment:
What does `gbvn` mean ?

Reply:
group by video names, now the code is updated